store/postgres/src/entities.rs

```diff
@@ -322,18 +332,35 @@ impl<'a> Connection<'a> {
         }
     }
 
-    /// Return the storage for the subgraph
-    fn storage(&self, subgraph: &SubgraphDeploymentId) -> Result<Storage, StoreError> {
+    /// Return the storage for the subgraph. Since constructing a `Storage`
+    /// object takes a bit of computation, we cache it at two levels:
+    /// all storage objects are cached within the connection and therefore for
+    /// the duration of a single transaction. Storage objects that do not have
+    /// a pending migration are also cached in the Store, i.e., for the
+    /// lifetime of the Store. Storage objects with a pending migration can
+    /// not be cached for longer than a transaction since they might change
+    /// without us knowing
+    fn storage(&self, subgraph: &SubgraphDeploymentId) -> Result<Arc<Storage>, StoreError> {
         let mut cache = self.cache.borrow_mut();
 
-        match cache.get(subgraph) {
-            Some(storage) => Ok(storage.clone()),
-            None => {
-                let storage = Storage::new(&self.conn, subgraph, self.store)?;
-                cache.insert(subgraph.clone(), storage.clone());
-                Ok(storage)
-            }
+        if let Some(storage) = cache.get(subgraph) {
```

Comment:
Is it not overkill to keep two caches, the connection and the store one?

Reply:
There's two reasons why I did it this way:

1. In a txn with a lot of operations, we do not need to get the Mutex every time (the per-connection cache is lockless)
2. For storage objects that have a migration pending, we can not put them into the `Store` cache (cross txn), but would still like to avoid recomputing the `Layout` for each operation in one txn

Reply:
Ok; taking out the per-txn cache